src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -2495,6 +2495,42 @@ void Lowering::ContainCheckSelect(GenTreeOp* node)
     {
         MakeSrcContained(node, op2);
     }
+
+    if (op1->IsCnsIntOrI() && op2->IsCnsIntOrI() && cond->OperIsSimple() &&
+        !varTypeIsFloating(cond->gtGetOp1()->TypeGet()))
+    {
+        assert(!varTypeIsFloating(cond->gtGetOp2()->TypeGet()));
+        ssize_t diff     = op1->AsIntCon()->IconValue() - op2->AsIntCon()->IconValue();
+        ssize_t maxValue = INT64_MAX;
+        ssize_t minValue = INT64_MIN;
+        if (op1->TypeGet() == TYP_INT)
+        {
+            diff     = (int)diff;
+            maxValue = INT32_MAX;
+            minValue = INT32_MIN;
+        }
+
+        if (abs(diff) == 1)
+        {
+            // When abs(op1 - op2) = 1, contain the larger op and emit cinc using the smaller op instead of csel with
+            // both the ops.
+            int op1Value = op1->AsIntCon()->IconValue();
+            int op2Value = op2->AsIntCon()->IconValue();
+            if (op1Value > op2Value || (op1Value == minValue && op2Value == maxValue))
+            {
+                MakeSrcContained(node, op1);
+                // Put smaller operand as op1 to simpilify checks in codegen
+                std::swap(node->gtOp1, node->gtOp2);
```

Comment:
I don't quite understand how this works. Doesn't this all of a sudden mean the semantics of `GT_SELECT` are changed to be the reverse of the normal case when one constant is contained?

Reply:
This is agreeably confusing because csel choses based on position (op1 or op2) while cinc increments value, op1 in our case.
When op1 > op2, the true case is higher so we can use the existing condition but not the position (the smaller value is in op1). Thus we use swap to bring it in op1.
When op2 > op1, the false case higher so we can use the existing positions (op1 has a smaller value) but not the condition. We want to increment in false case, i.e., when the reverse of the existing condition is true. Thus we only reverse the condition not swap the operands. 🙂

Reply:
What is the LIR produced after this case runs? Does it match the normal semantics for `GT_SELECT`? If something else ends up unmarking the operand as contained is the codegen then suddenly wrong?